scripts/lib/upgrade-zulip-stage-2

```diff
@@ -445,11 +449,39 @@ def run_hooks(kind: Literal["pre-deploy", "post-deploy"]) -> None:
     path = f"/etc/zulip/hooks/{kind}.d"
     if not os.path.exists(path):
         return
+    # Pass in, via environment variables, the old/new "version
+    # string" (which is a `git describe` output)
+    env = os.environ.copy()
+    env["ZULIP_OLD_VERSION"] = old_version
+    env["ZULIP_NEW_VERSION"] = NEW_ZULIP_VERSION
+
+    # preexec_fn=su_to_zulip normally handles this, but our explicit
+    # env overrides that
+    env["HOME"] = get_zulip_pwent().pw_dir
+
+    def resolve_version_string(version: str) -> str:
+        matches = re.search(r"-\d+-g([a-f0-9]{7,})$", version)
+        to_parse = version
+        if matches is not None:
+            to_parse = matches.group(1)
+        return subprocess.check_output(
+            ["git", "rev-parse", to_parse], cwd=deploy_path, preexec_fn=su_to_zulip, text=True
+        ).strip()
```

Comment:
There is no guarantee that `git describe` strings look like that; for example, `7.0` or `7.0-beta3` are just as possible as `7.0-dev-2022-g18d8fbc74b`.

Also, we should never need to rip these strings apart with regexes, because Git already understands them.

```console
$ git rev-parse 7.0-dev-2022-g18d8fbc74b
18d8fbc74be15ac20ee74433abe3263d8e8b7208
```

Also, it’s evidently possible for this to fail when the initial installation was from a Git commit that isn’t upstream.

```
2023-06-14 04:02:05,454 upgrade-zulip-stage-2: Pre-checking for puppet changes...
fatal: ambiguous argument '870443cd0b': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
Traceback (most recent call last):
  File "/home/zulip/deployments/2023-06-14-03-59-44/scripts/lib/upgrade-zulip-stage-2", line 518, in <module>
    run_hooks("pre-deploy")
  File "/home/zulip/deployments/2023-06-14-03-59-44/scripts/lib/upgrade-zulip-stage-2", line 497, in run_hooks
    env["ZULIP_OLD_COMMIT"] = resolve_version_string(old_version)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/zulip/deployments/2023-06-14-03-59-44/scripts/lib/upgrade-zulip-stage-2", line 489, in resolve_version_string
    return subprocess.check_output(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['git', 'rev-parse', '870443cd0b']' returned non-zero exit status 128.

Zulip upgrade failed (exit code 1)!

The upgrade process is designed to be idempotent, so you can retry after resolving whatever issue caused the failure (there should be a traceback above). A log of this installation is available in /var/log/zulip/upgrade.log
```

Reply:
> There is no guarantee that `git describe` strings look like that; for example, `7.0` or `7.0-beta3` are just as possible as `7.0-dev-2022-g18d8fbc74b`.

That's why if it fails to match the regex, we `rev-parse` the whole string.

> Also, we should never need to rip these strings apart with regexes, because Git already understands them.

TIL!  I feel like it didn't when I first learned it -- either I missed it, or I've not re-read the help since git/git@7dd45e15c259e44b1c8b5ffdfc0c3d002c7f642c.  I'll drop the three lines before the rev-parse.

> Also, it’s evidently possible for this to fail when the initial installation was from a Git commit that isn’t upstream.

I think this happens if you install from git from something not `upstream/main` -- since that has a `~root/zulip/.git/` and those objects and refs don't get into `/srv/zulip.git` when `upgrade-zulip-from-git` runs next.

I think the fix here is to, when we move the install dir into `~zulip/deployments/current`, also copy/move the `.git` to `/srv/zulip.git`, mark it bare, and maybe set `~zulip/deployments/current` as a worktree.

Reply:
Pushed #26025.